src/sessions/softphone-session-handler.ts

```diff
@@ -22,22 +96,319 @@ export default class SoftphoneSessionHandler extends BaseSessionHandler {
     }
   }
 
+  handleSoftphoneConversationUpdate (
+    update: ConversationUpdate,
+    participant: IConversationParticipantFromEvent,
+    callState: ICallStateFromParticipant,
+    session?: IExtendedMediaSession
+  ) {
+    const conversationId = update.id;
+    const lastConversationUpdate = this.conversations[conversationId] ? { ...this.conversations[conversationId] } : null;
+
+    if (session) {
+      session.pcParticipant = participant as any;
+    }
+
+    this.log('debug', 'about to process conversation event', { session, update, lastConversationUpdate, callState });
+
+    /* if we didn't have a previous update and this one is NOT in a pending ended state, that means we are not responsible for this conversation (another client handled it or we have already emitted the `sessionEnded` for it) */
+    if (!lastConversationUpdate && !this.isPendingState(callState)) {
+      return this.log('debug', 'received a conversation event for a conversation we are not responsible for. not processing', { update, callState }, true);
+    }
+
+    /* update the conversation state */
+    if (!lastConversationUpdate) {
+      this.conversations[conversationId] = {} as any;
+    }
+
+    /* this is just a precaution – session for the conversation should be added in `this.acceptSession()` */
+    if (!this.conversations[conversationId].session) {
+      this.conversations[conversationId].session = session;
+    }
+
+    this.conversations[conversationId].mostRecentCallState = callState;
+    this.conversations[conversationId].mostRecentUserParticipant = participant;
+    this.conversations[conversationId].conversationUpdate = update;
+    this.conversations[conversationId].conversationId = conversationId;
+
+    const previousCallState = this.getUsersCallStateFromConversationEvent(lastConversationUpdate?.conversationUpdate);
+
+    /* if the state didn't change at all, we can skip processing */
+    if (!this.diffConversationCallStates(previousCallState, callState)) {
+      return this.log('debug', 'conversation update received but state is unchanged. ignoring', {
+        conversationId,
+        previousCallState,
+        callState,
+        sessionType: this.sessionType
+      });
+    }
+
+    const communicationStateChanged = previousCallState?.state !== callState.state;
```

Comment:
NIT: use `const communicationStateChange = this.diffConversationCallStates(previousCallState, callState)` and use the same variable in your if-check.

Reply:
The reason I'm not using `diffConversationCallStates()` is because that take into account more factors (like muted, held, etc). This check is looking at only the communication state (ie. are we going from `pending` to `active`, etc).